### PR TITLE
functoria: move packages functions into a Functoria_package module.

### PR DESCRIPTION
**Pull request by samoht**
This also abstracts the Functoria.package type, and rename the constructor
and accessors to use more sensible names.

We also vendor the function to compare opam version to simplify the code.

### CODE REVIEW

File: functoria/lib/functoria_package_version.ml [hunk 1]
```diff
@@ -0,0 +1,168 @@
+(******************************************************************************)
+(*  This file is part of the Dose library http://www.irill.org/software/dose  *)
+(*                                                                            *)
+(*  Copyright (C) 2011 Ralf Treinen <ralf.treinen@pps.jussieu.fr>             *)
+(*                                                                            *)
+(*  All rights reserved. This file is distributed under the terms of the      *)
+(*  GNU Lesser General Public License version 2.1, with the special           *)
+(*  exception on linking described in the file LICENSE.                       *)
```
Comment (hannesm) [inline functoria/lib/functoria_package_version.ml]: I'm not sure whether we want to vendor LGPL code in mirage..

Comment (hannesm) [reply functoria/lib/functoria_package_version.ml]: and with recent functoria changes, we shouldn't need to compare version numbers... https://github.com/mirage/functoria/pull/183

Comment (samoht) [reply functoria/lib/functoria_package_version.ml]: hum ok I'll revert this change. I was not totally sure about the current code, but let's keep it as is :-)

Comment (hannesm) [reply functoria/lib/functoria_package_version.ml]: sorry for failing to document the current code. the main idea is to track the versions as strings around, and emit whatever constraints are there to the opam file. then opam will figure out whether there's a solution with all the constraints (and we don't need to handle version comparison).

Comment (samoht) [reply functoria/lib/functoria_package_version.ml]: makes sense, I've reverted my change and updated the tests to `Package.merge` to show what happens after a merge for min/max versions.

**Comment by Drup** [comment]
Looks good!

**Pull request merged**